glossary (#34)
graphql/composite-schemas-wg: docs/glossary.md

schema type definitions.

### Apollo Federation Subgraph ([Link](https://www.apollographql.com/docs/federation/subgraphs/))

A service that exposes a GraphQL schema that is consumed by a gateway in order
to create a federated graph. The services schema uses directive annotations on
types and fields as a hint for a gateway on how it should combine the services
schema with the schema of the other Apollo federation subgraphs. A Apollo
Federation Subgraph is a GraphQL schema with Apollo Federation schema
directives. Also See **Subschema**.

### Apollo Graph Platform

Branding used by Apollo for describing their open-source products.

> The leading open-source GraphQL implementation with 17M downloads per month
> and the only end-to-end cloud management solution for GraphQL. Source:
> [https://www.apollographql.com/](https://www.apollographql.com/) (2022/02/24)

### Apollo Studio

A schema registry for registering and validating Apollo federation subgraphs.

### Data Source

A “dumb” service or database that exposes raw data. Often a data source is
consumed by a service, for reading and/or persisting business logic state.

> It is hard to define a border between Data Source and Service. E.g. a PostgreSQL

Reviewer: Isn't the difference between Source and Service the presence of business logic?

Reviewer 2: Maybe _data source_ is not the best term here. A data source does not directly refer to the persistence layer of data, so I can source data from a DB via an API/Interface that also include embedded business logic; the data source is the API, not the DB. 
Nonetheless, it's also true that a view on a database can include a business logic. A query can include business logic as well, unless it's a `select * from ...`. The example of the trigger is quite correct.

into a GraphQL schema. In the case of multiple data sources or services, Mesh
uses Schema Stitching for doing Schema Composition of the GraphQL subschemas
that were the results of the transform.

### Hybrid Schema Composition

A mix of both **Subschema Directive-based Schema Composition** and
**Programmatic/Config Gateway-based Schema Composition**.

### Nadel ([Link](https://github.com/atlassian-labs/nadel))

> Nadel is a Kotlin library to combine several GraphQL services together. This
> is achieved by combining several underlying GraphQL services (schemas) into
> one schema (overall schema). The main purpose of this library is to power a
> GraphQL gateway which serves as the entry point for executing requests across
> different services while presenting one GraphQL API.

### Nautilus ([Link](https://github.com/nautilus/gateway))

> A library and standalone service that composes your GraphQL APIs into one
> endpoint.

Uses these
[merge strategies](https://github.com/nautilus/gateway/blob/master/docs/mergingStrategies.md)
and one interface type to compose subschemas (see **Node interface**).

### Node interface

> The [Global Object
> Identification][https://graphql.org/learn/global-object-identification/]
> [spec][https://relay.dev/graphql/objectidentification.htm] defines a `Node`

Reviewer 3: Links fixes
```suggestion
The [Global Object
Identification](https://graphql.org/learn/global-object-identification/)
[spec](https://relay.dev/graphql/objectidentification.htm) defines a `Node`
```

Alternatively, you can use the `[words][link-ref-name]` followed by `[link-ref-name]: https://...` syntax like below 🤷

### Apollo Federated Graph ([Link](https://www.apollographql.com/docs/federation/))

The process of composing multiple Apollo Subgraphs into one schema, usually via a Gateway.

### Apollo Studio

A schema registry for registering and validating Apollo federation subgraphs.

### Apollo Federation Gateway

A gateway that usually uses a schema registry (Apollo Studio) for fetching the apollo subgraph configurations. It is the entry point for incoming GraphQL requests. It analyzes the incoming operations, builds a query plan in order to find the quickest way of retrieving all the data from the apollo federation subgraphs, and then executes that plan.

The approach of combining multiple GraphQL services into a unified GraphQL Graph on a Gateway using meta-information applied on the individual GraphQL services schema definitions.

> 

Reviewer 3: Thanks for putting this together @n1ru4l 👏 

I'm not 100% sure if this is covered by an existing glossary item. I think it's unique in that is isn't specifying directives and only requires the `id: ID!` field is implemented via interface.

The spec for "global object identification" lends itself well to schema merging for my team's needs. This solved many of our schema composition problems in a very simple way, i.e. subschemas did not need to go out of their way to opt-in to composition.
```suggestion

### Node interface

The [Global Object Identification][] [spec][relay-goi-spec] defines a `Node` interface, which enables composition if shared types implement it.

[Nautilus Gateway](https://github.com/nautilus/gateway) is one such implementation. It uses these [merge strategies](https://github.com/nautilus/gateway/blob/master/docs/mergingStrategies.md) and one interface type to compose subschemas.

[Global Object Identification]: https://graphql.org/learn/global-object-identification/
[relay-goi-spec]: (https://relay.dev/graphql/objectidentification.htm)
```

Author: done via https://github.com/graphql/composite-schemas-wg/pull/34/commits/a511f6fc2469ba8a6f68b623aedf5239e2da84db

### Subschema

> A  Subschema is a GraphQL schema that is used in a type merging contexts, such as in Apollo federation or Schema Stitching for. In Apollo federation, a Subschema is often called a Apollo Federation Subgraph.

Author: ```suggestion
A  Subschema is a GraphQL schema that is used in schema composition contexts, such as in Apollo federation or Schema Stitching. In Apollo federation, a Subschema is often called an Apollo Federation Subgraph.
```

Reviewer 2: I would also add a link for [Apollo Federation Subgraph](https://www.apollographql.com/docs/federation/federation-spec/)

Author: done, thanks!

### Subschema Directive-based Type Merging

> The process of applying type merging on the Gateway using the directive meta-information exposed on the consumed GraphQL services (**Subschemas** as done by **Apollo Federation** or **Schema Stitching**.

Author: ```suggestion
The process of applying schema composition on the Gateway using the directive meta-information exposed on the consumed GraphQL services (**Subschemas** as done by **Apollo Federation** or **Schema Stitching**.
```

> ### Programmatic/Config Gateway-based Type Merging

Author: ```suggestion
### Programmatic/Config Gateway-based Schema Composition
```

A method of Type Merging where the **Subschemas** do not expose any information on how they can be merged. Instead, the rules are defined on the Gateway, either as code or a configuration file.

### Hybrid Type Merging

> A mix of both **Subschema Directive-based Type Merging** and **Programmatic/Config Gateway-based Type Merging**.

Author: ```suggestion
A mix of both **Subschema Directive-based Schema Composition** and **Programmatic/Config Gateway-based Schema Composition**.
```
